cmd/compute-domain-kubelet-plugin/driver.go

```diff
@@ -171,73 +173,75 @@ func (d *driver) NodeUnprepareResources(ctx context.Context, req *drapbv1.NodeUn
 
 	workQueue.Run(ctx)
 
-	return unpreparedResources, nil
+	return results, nil
 }
 
-func (d *driver) nodePrepareResource(ctx context.Context, claim *drapbv1.Claim) (bool, *drapbv1.NodePrepareResourceResponse) {
+func (d *driver) nodePrepareResource(ctx context.Context, claim *resourceapi.ResourceClaim) (bool, kubeletplugin.PrepareResult) {
 	d.Lock()
 	defer d.Unlock()
 
-	resourceClaim, err := d.client.ResourceV1beta1().ResourceClaims(claim.Namespace).Get(
-		ctx,
-		claim.Name,
-		metav1.GetOptions{})
-	if err != nil {
-		ret := &drapbv1.NodePrepareResourceResponse{
-			Error: fmt.Sprintf("failed to fetch ResourceClaim %s in namespace %s", claim.Name, claim.Namespace),
-		}
-		return isPermanentError(err), ret
-	}
-
-	if resourceClaim.Status.Allocation == nil {
-		ret := &drapbv1.NodePrepareResourceResponse{
-			Error: fmt.Sprintf("no allocation set in ResourceClaim %s in namespace %s", claim.Name, claim.Namespace),
+	if claim.Status.Allocation == nil {
+		res := kubeletplugin.PrepareResult{
+			Err: fmt.Errorf("no allocation set in ResourceClaim %s in namespace %s", claim.Name, claim.Namespace),
 		}
-		return true, ret
+		return true, res
 	}
 
-	prepared, err := d.state.Prepare(ctx, resourceClaim)
+	devs, err := d.state.Prepare(ctx, claim)
 	if err != nil {
-		ret := &drapbv1.NodePrepareResourceResponse{
-			Error: fmt.Sprintf("error preparing devices for claim %v: %v", claim.UID, err),
+		res := kubeletplugin.PrepareResult{
+			Err: fmt.Errorf("error preparing devices for claim %v: %v", claim.UID, err),
 		}
-		return isPermanentError(err), ret
+		return isPermanentError(err), res
+	}
+
+	// Translate type: drapbv1.Device -> kubeletplugin.Device
+	// Maybe instead change return type of state.Prepare()
+	var prepDevs []kubeletplugin.Device
+	for _, d := range devs {
+		device := kubeletplugin.Device{
+			// kubeletplugin/draplugin.go has a stripSubrequestNames() helper:
+			// Requests: stripSubrequestNames(d.RequestNames),
+			Requests:     d.RequestNames,
+			PoolName:     d.PoolName,
+			DeviceName:   d.DeviceName,
+			CDIDeviceIDs: d.CDIDeviceIDs,
+		}
+		prepDevs = append(prepDevs, device)
 	}
 
-	klog.Infof("Returning newly prepared devices for claim '%v': %v", claim.UID, prepared)
-	return true, &drapbv1.NodePrepareResourceResponse{Devices: prepared}
+	klog.Infof("Returning newly prepared devices for claim '%v': %v", claim.UID, prepDevs)
+	return true, kubeletplugin.PrepareResult{Devices: prepDevs}
 }
 
-func (d *driver) nodeUnprepareResource(ctx context.Context, claim *drapbv1.Claim) (bool, *drapbv1.NodeUnprepareResourceResponse) {
+func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugin.NamespacedObject) (bool, error) {
 	d.Lock()
 	defer d.Unlock()
 
-	resourceClaim, err := d.client.ResourceV1beta1().ResourceClaims(claim.Namespace).Get(
+	claim, err := d.client.ResourceV1beta1().ResourceClaims(claimNs.Namespace).Get(
```

Comment:
interesting asymmetry: while the new helper injects full claim objects into `PrepareResourceClaims()` it does not do so for `UnPrepareResourceClaims()`, i.e. here we still need to GET them 'manually'.

Reply:
@pohly is there a reason for this asymmetry?

Reply:
Yes. Unpreparing has to work also in scenarios where the ResourceClaim is gone. This is the "determined user force-deletes claim and pod while it is running" scenario. The kubelet then still needs to be able to tear down the pod.

Reply:
Let's fix this in a follow-up